step1 entity 매핑

yevgnenll opened the pull request:
안녕하세요

너무 늦게 코드리뷰 요청 드려서 죄송합니다 ㅠㅠ
야근이 밀려와서 이제서야 하고 있습니다 최대한 빠르게 진행하겠습니다

Commit:
test: answer entity 가 저장되는지 검증한다

저장된다면 id, created_at 이 할당된다

Commit:
chore: 설정 변경

1. mode: mysql
2. show-sql 로 실행되는 sql 을 보여준다
3. sql query 는 formatting 을 거친다

Commit:
feat: 첫 번째 테스트가 통과하도록 entity 설정

빈이 로드 되면서 jpa 에 관련된 모든 entity 의 설정이
필요하여, 먼저 통과되는 테스트를 위해 필요한 설정을 수행한다

Commit:
feat: uniqueConstraint 설정

user 스키마에 명시된 설정 추가

Commit:
test: select 가 되는지 검증한다

Commit:
refactor: 날짜 생성방식 변경

1. annotation 을 사용하는 것이 더 깔끔하다 판단하여 수정
2. updated_at 은 setter 가 존재하면 안되다고 판단하여 제거

Commit:
fix: 요구사항대로 column type 변경

Commit:
test: update 실행시 update_at 갱신을 검증한다

Commit:
chore: 테이블 생성 쿼리 주석으로 추가

Commit:
test: DeleteHistory 에 대한 테스트 작성

insert 와 select 에 대한 기능을 검증한다

Commit:
feat: content column type 정의, 날짜 추가

Commit:
test: question 에 대한 쿼리 검증

repository 에 정의된 쿼리들을 검증한다.
그 외에 insert, select, update 를 검증한다

Commit:
test: updated_at, created_at 생성 검증

Commit:
feat: 컬럼 매핑, index 추가

Commit:
test: 정의된 public 함수들 검증 및 repository 검증

Commit:
docs: 요구사항 추가

kingbbode commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,52 @@
 public class AnswerTest {
-    public static final Answer A1 = new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
-    public static final Answer A2 = new Answer(UserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
+	public static final Answer A1 = new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, "Answers Contents1");
+	public static final Answer A2 = new Answer(UserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
```
JPA 를 사용할 때에는 객체의 영속 상태를 조심스럽게 다루어야 합니다. 이런 공유객체를 사용하면서 `DataJpaTest` 를 사용하게 되었을 때 발생하는 문제점이 많습니다 :)

전역으로 선언된 이 객체가 테스트가 진행되는 동안 언제 어떻게 영속 상태의 변화를 받는지도 알고 계시다면 좋겠지요.

---

모든 테스트는 상호 독립적으로 작성되어야 하는 것이 테스트의 필수 조건이기도 합니다. `A1`, `A2` 는 JPA 특성상 단순 객체를 넘어서 모든 테스트에 공유가 되고 있습니다.

각 테스트의 조건은 각 테스트 안에서만 동작하도록 변경 부탁드립니다 :)

yevgnenll replied:
`@Transactional` 때문에 상관 없을거라고 생각했는데 그렇지가 않네요...
이건 회사 코드에서도 리팩토링 해야할것 같습니다.
전역 객체를 저장해서 사용하는 코드는 모두 제외 하였습니다 :)

kingbbode commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,52 @@
+
+	@BeforeEach
+	void init() {
+		answerRepository.save(A2);
```
테스트의 조건은 테스트 안에서 작성이 되어야합니다.
테스트의 setup 단계는 테스트를 위한 환경을 세팅하는 단계입니다.

이 테스트가 조건인지 환경인지가 구분이 어려울 수도 있는데요.
setup 에서 설치된 환경이 테스트 안에서 변경될 수 있다면 그것은 테스트의 조건일 가능성이 높습니다.

작성하신 setup의 내용은 테스트의 조건에 해당하는 것으로 보입니다 :

yevgnenll replied:
넵 테스트에 대한 조건이 맞습니다ㅠㅠ
테스트 안으로 해당하는 조건을 옮기도록 하겠습니다~!!

kingbbode commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,40 @@
+    @Column(name = "question_id")
     private Long questionId;
+
+    @Column(columnDefinition = "clob")
```
`@Lob`  을 사용하지 않는 이유는 무엇인가요? :)

yevgnenll replied:
사실 `@Lob`을 까먹고 있다가 노션에 정리해둔걸 보고 Question 에만 적용하고 Answer 에는 누락했습니다 ㅠ

kingbbode reviewed (changes requested):
코멘트 작성해두었습니다 :)
확인 부탁드립니다.

Commit:
refactor: Lob 사용

Commit:
test: 테스트 케이스 분리, 선제 조건 격리

1. 전역 객체를 영속성 안으로 들어가지 않도록 변경
2. 테스트에 필요한 선제 조건은 각 테스트에서 실행하도록 개선

Commit:
test: 코드리뷰 반영

1. test 선제 조건을 각 테스트 코드에서 실행하여 격리한다
2. 전역 객체가 영속성을 갖도록 하여 다른 테스트에 영향을 끼치지 않도록
전역 객체를 테스트에서 사용하지 않는다

Commit:
chore: 쿼리를 보는 옵션과 table 을 생성하고 drop 하는 구성 추가

kingbbode commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,72 @@
+
+@DataJpaTest
 public class QuestionTest {
     public static final Question Q1 = new Question("title1", "contents1").writeBy(UserTest.JAVAJIGI);
     public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
```
모두 제거었다면 더 좋았을 것 같네요 :)

kingbbode reviewed (approved):
잘 구현하셨네요 :)
이만 머지하겠습니다.
